README.md
# go-crond

[![GitHub release](https://img.shields.io/github/release/webdevops/go-crond.svg)](https://github.com/webdevops/go-crond/releases)
[![license](https://img.shields.io/github/license/webdevops/go-crond.svg)](https://github.com/webdevops/go-crond/blob/master/LICENSE)
[![Build Status](https://travis-ci.org/webdevops/go-crond.svg?branch=master)](https://travis-ci.org/webdevops/go-crond)
[![Github All Releases](https://img.shields.io/github/downloads/webdevops/go-crond/total.svg)]()
[![Github Releases](https://img.shields.io/github/downloads/webdevops/go-crond/latest/total.svg)]()

A cron daemon written in golang

Inspired by https://github.com/anarcher/go-cron

Using https://godoc.org/github.com/robfig/cron

## Docker images

on [Docker hub](https://hub.docker.com/repository/docker/webdevops/go-crond/tags)

- `webdevops/go-crond:alpine` (based on `alpine`)
- `webdevops/go-crond:ubuntu` (based on `ubuntu:latest`)
- `webdevops/go-crond:debian` (based on `debian:stable-slim`)
- `webdevops/go-crond:{version}-alpine` (based on `alpine`)
- `webdevops/go-crond:{version}-ubuntu` (based on `ubuntu:latest`)
- `webdevops/go-crond:{version}-debian` (based on `debian:stable-slim`)

## Features

- system crontab (with username inside)
- user crontabs (without username inside)
- run-parts support
- Logging to STDOUT and STDERR (instead of sending mails)
- Keep current environment (eg. for usage in Docker containers)
- Supports Linux, MacOS, ARM/ARM64 (Rasbperry Pi and others)

## Usage

```
Usage:
  go-crond [user:crontab] [user:crontab]…

Application Options:
      --default-user=       Default user (default: root)
      --include=            Include files in directory as system crontabs (with user)
      --auto                Enable automatic system crontab detection
      --run-parts=          Execute files in directory with custom spec (like run-parts; spec-units:ns,us,s,m,h; format:time-spec:path; eg:10s,1m,1h30m)
      --run-parts-1min=     Execute files in directory every beginning minute (like run-parts)
      --run-parts-15min=    Execute files in directory every beginning 15 minutes (like run-parts)
      --run-parts-hourly=   Execute files in directory every beginning hour (like run-parts)
      --run-parts-daily=    Execute files in directory every beginning day (like run-parts)
      --run-parts-weekly=   Execute files in directory every beginning week (like run-parts)
      --run-parts-monthly=  Execute files in directory every beginning month (like run-parts)
      --allow-unprivileged  Allow daemon to run as non root (unprivileged) user
  -V, --version             show version and exit
      --dumpversion         show only version number and exit
  -h, --help                show this help message
  -v, --verbose             verbose mode [$VERBOSE]
      --log.json            Switch log output to json format [$LOG_JSON]
      --server.bind=        Server address, eg. ':8080' (/healthz and /metrics for prometheus) [$SERVER_BIND]
      --server.metrics      Enable prometheus metrics (do not use senstive informations in commands -> use environment variables or files for storing these informations) [$SERVER_METRICS]
```

Crontab files can be added as arguments or automatic included by using eg. `--include=crond-path/`

### Examples

Run crond with a system crontab:

    go-crond examples/crontab


Run crond with user crontabs (without user in it) under specific users:

    go-crond \
        root:examples/crontab-root \ 
        guest:examples/crontab-guest


Run crond with auto include of /etc/cron.d and script execution of hourly, weekly, daily and monthly:

    go-crond \
        --include=/etc/cron.d \
        --run-parts-hourly=/etc/cron.hourly \
        --run-parts-weekly=/etc/cron.weekly \
        --run-parts-daily=/etc/cron.daily \
        --run-parts-monthly=/etc/cron.monthly

Run crond with run-parts with custom time spec:

    go-crond \
        --run-parts=1m:/etc/cron.minute \
        --run-parts=15m:/etc/cron.15min

Run crond with run-parts with custom time spec and different user:

    go-crond \
        --run-parts=1m:application:/etc/cron.minute \
        --run-parts=15m:admin:/etc/cron.15min

## Installation

```bash
GOCROND_VERSION=20.7.0 \
&& wget -O /usr/local/bin/go-crond https://github.com/webdevops/go-crond/releases/download/$GOCROND_VERSION/go-crond-64-linux \
&& chmod +x /usr/local/bin/go-crond
```

## Docker images

| Image                       | Description                                    |
|:----------------------------|:-----------------------------------------------|
| `webdevops/go-crond:latest` | Latest release, binary only                    |
| `webdevops/go-crond:master` | Current development version in branch `master` |

## Metrics

go-crond exposes [Prometheus][] metrics on `:8080/metrics` if enabled.


| Metric                      | Description                                     |
|:----------------------------|:------------------------------------------------|
| `gocrond_task_info`         | List of all cronjobs                            |
| `gocrond_task_run_count`    | Counter for each executed task                  |
| `gocrond_task_run_success`  | Last status (0=failed, 1=success) for each task |
| `gocrond_task_run_time`     | Last exec time (unix timestamp) for each task   |
| `gocrond_task_run_duration` | Duration of last exec                           |

[Prometheus]: https://prometheus.io/


CHANGELOG.md
# Change Log

## [20.7.0] - 2020-07-01
- Disable http server by default
- Cleanup
- Added env variable `LOG_JSON` and `VERBOSE`
- Added various Docker image builds (alpine, debian, ubuntu)
- Switch to `ENTRYPOINT` instead of `CMD` for Dockerfiles

## [20.6.0] - 2020-06-29
- Switch to [Calendar Versioning](https://calver.org/) with `YY.MM.MICRO format` (year, month, increment)
- Implemented new build system
    - inject version info from git
    - linting
- Add prometheus metric support
- Switch logger to logrus
- Replace `--no-auto` with `--auto` (system crontab needs to be enabled now)

## [0.6.0] - 2017-06-01
- Switching of current working directory to / (root) when running cronjobs
- Add cronjob elapsed time in log
- Improve distribution detection and automatic fallback
- Switch from dynamic to static linking by default (dynamic is still available with `-dynamic` suffix)
- Replaced and reversed ` --system-defaults` to `--no-auto` (will now disable including of system default crontabs)
- Remove `--threads` (use env var `GOMAXPROCS` instead)

## [0.5.1] - 2017-05-26
- Fix crosscompiled binaries

## [0.5.0] - 2017-05-25
- Add daemon reload on SIGHUP
- Add `--allow-unprivileged` for running without root rights and without user switching capability

## [0.4.0] - 2017-05-13
- Replace `--processes` with `--threads`
- Fix argument parsing (segfault if invalid argument/option is passed)
- Include `/etc/cron.d/` as system defaults
- Improve logging

## [0.3.0] - 2017-04-28
- Add `--dumpversion`
- Add user support for `--run-parts`

## [0.2.0] - 2017-04-25
*Development release*

## [0.1.0] - 2017-04-25
*Development release*


cron_test.go
package main

import (
	"testing"
)

func TestReadCronFile(t *testing.T) {
	path := "./crontab"

	c, err := NewCron(path)

	if err != nil {
		t.Error(err)
	}

	if c == nil {
		t.Error("cron is empty")
	}

}


parser_test.go
package main

import (
	"regexp"
	"strings"
	"testing"
)

func TestParser(t *testing.T) {

	s := `
# Test
*/10 * * * * echo "hello world"
* */1 * * * echo "hello world"
`
	strReader := strings.NewReader(s)
	p, err := NewParser(strReader)

	if err != nil {
		t.Error(err)
	}

	if p == nil {
		t.Error("parser is nil!")
	}

}

func TestParserRegex(t *testing.T) {
	rp := regexp.MustCompile(LINE_RE)
	s := `*/10 * * * * echo "hello world"`
	f := rp.MatchString(s)
	if f == false {
		t.Error("not match")
		return
	}

	r := rp.FindStringSubmatch(s)
	t.Logf("%#v", r)

}


funcs.go
package main

import (
	"fmt"
	log "github.com/sirupsen/logrus"
	"os"
	"path/filepath"
	"strconv"
	"syscall"
)

func fileGetAbsolutePath(path string) (string, os.FileInfo) {
	ret, err := filepath.Abs(path)
	if err != nil {
		log.Fatalf("invalid file: %v", err)
	}

	f, err := os.Lstat(ret)
	if err != nil {
		log.Fatalf("file stats failed: %v", err)
	}

	return ret, f
}

func checkIfFileExists(path string) bool {
	f, err := os.Stat(path)

	// check if path exists
	if err != nil {
		return false
	}

	return (!f.IsDir()) && f.Mode().IsRegular()
}

func checkIfDirectoryExists(path string) bool {
	f, err := os.Stat(path)

	// check if path exists
	if err != nil {
		return false
	}

	return f.IsDir()
}

func checkIfFileExistsAndOwnedByRoot(path string) bool {
	f, err := os.Stat(path)

	// check if path exists
	if err != nil {
		return false
	}

	// check if it is not a file
	if !f.Mode().IsRegular() {
		return false
	}

	uidS := fmt.Sprint(f.Sys().(*syscall.Stat_t).Uid)
	uid, err := strconv.Atoi(uidS)
	if err != nil {
		return false
	}

	if uid != 0 {
		return false
	}

	return true
}

func checkIfFileIsValid(f os.FileInfo, path string) bool {
	if f.IsDir() {
		return false
	}

	if f.Mode().IsRegular() {
		if f.Mode().Perm()&0022 == 0 {
			return true
		} else {
			log.Infof("ignoring file with wrong modes (not xx22) %s\n", path)
		}
	} else {
		log.Infof("ignoring non regular file %s\n", path)
	}

	return false
}


parser.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"regexp"
	"strings"
)

const (
	ENV_LINE = `^(\S+)=(\S+)\s*$`

	//                     ----spec------------------------------------    --user--  -cmd-
	CRONJOB_SYSTEM = `^\s*([^@\s]+\s+\S+\s+\S+\s+\S+\s+\S+|@every\s+\S+)\s+([^\s]+)\s+(.+)$`

	//                  ----spec------------------------------------    -cmd-
	CRONJOB_USER = `^\s*([^@\s]+\s+\S+\s+\S+\s+\S+\s+\S+|@every\s+\S+)\s+(.+)$`

	DEFAULT_SHELL = "sh"
)

var (
	envLineRegex       = regexp.MustCompile(ENV_LINE)
	cronjobSystemRegex = regexp.MustCompile(CRONJOB_SYSTEM)
	cronjobUserRegex   = regexp.MustCompile(CRONJOB_USER)
)

type CrontabEntry struct {
	Spec    string
	User    string
	Command string
	Env     []string
	Shell   string
}

type Parser struct {
	cronLineRegex   *regexp.Regexp
	reader          io.Reader
	cronjobUsername string
}

// Create new crontab parser (user crontab without user specification)
func NewCronjobUserParser(reader io.Reader, username string) (*Parser, error) {
	p := &Parser{
		cronLineRegex:   cronjobUserRegex,
		reader:          reader,
		cronjobUsername: username,
	}

	return p, nil
}

// Create new crontab parser (crontab with user specification)
func NewCronjobSystemParser(reader io.Reader) (*Parser, error) {
	p := &Parser{
		cronLineRegex:   cronjobSystemRegex,
		reader:          reader,
		cronjobUsername: CRONTAB_TYPE_SYSTEM,
	}

	return p, nil
}

// Parse crontab
func (p *Parser) Parse() []CrontabEntry {
	entries := p.parseLines()

	return entries
}

// Parse lines from crontab
func (p *Parser) parseLines() []CrontabEntry {
	var entries []CrontabEntry
	var crontabSpec string
	var crontabUser string
	var crontabCommand string
	var environment []string

	shell := DEFAULT_SHELL

	specCleanupRegexp := regexp.MustCompile(`\s+`)

	scanner := bufio.NewScanner(p.reader)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())

		// comment line
		if strings.HasPrefix(line, "#") {
			continue
		}

		// environment line
		if envLineRegex.MatchString(line) == true {
			m := envLineRegex.FindStringSubmatch(line)
			envName := strings.TrimSpace(m[1])
			envValue := strings.TrimSpace(m[2])

			if envName == "SHELL" {
				// custom shell for command
				shell = envValue
			} else {
				// normal environment variable
				environment = append(environment, fmt.Sprintf("%s=%s", envName, envValue))
			}
		}

		// cronjob line
		if p.cronLineRegex.MatchString(line) == true {
			m := p.cronLineRegex.FindStringSubmatch(line)

			if p.cronjobUsername == CRONTAB_TYPE_SYSTEM {
				crontabSpec = strings.TrimSpace(m[1])
				crontabUser = strings.TrimSpace(m[2])
				crontabCommand = strings.TrimSpace(m[3])
			} else {
				crontabSpec = strings.TrimSpace(m[1])
				crontabUser = p.cronjobUsername
				crontabCommand = strings.TrimSpace(m[2])
			}

			// shrink white spaces for better handling
			crontabSpec = specCleanupRegexp.ReplaceAllString(crontabSpec, " ")

			entries = append(entries, CrontabEntry{Spec: crontabSpec, User: crontabUser, Command: crontabCommand, Env: environment, Shell: shell})
		}
	}

	return entries
}


runner.go
package main

import (
	"fmt"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/robfig/cron"
	log "github.com/sirupsen/logrus"
	"os"
	"os/exec"
	"os/user"
	"strconv"
	"strings"
	"syscall"
	"time"
)

type Runner struct {
	cron *cron.Cron
}

func NewRunner() *Runner {
	r := &Runner{
		cron: cron.New(),
	}
	return r
}

// Add crontab entry
func (r *Runner) Add(cronjob CrontabEntry) error {
	cronSpec := cronjob.Spec
	if !strings.HasPrefix(cronjob.Spec, "@") {
		cronSpec = fmt.Sprintf("0 %s", cronjob.Spec)
	}

	err := r.cron.AddFunc(cronSpec, r.cmdFunc(cronjob, func(execCmd *exec.Cmd) bool {
		// before exec callback
		log.WithFields(LogCronjobToFields(cronjob)).Debugf("executing")
		return true
	}))

	if err != nil {
		prometheusMetricTask.With(r.cronjobToPrometheusLabels(cronjob)).Set(0)
		log.WithFields(LogCronjobToFields(cronjob)).Errorf("cronjob failed adding:%v", err)
	} else {
		prometheusMetricTask.With(r.cronjobToPrometheusLabels(cronjob)).Set(1)
		log.WithFields(LogCronjobToFields(cronjob)).Infof("cronjob added")
	}

	return err
}

// Add crontab entry with user
func (r *Runner) AddWithUser(cronjob CrontabEntry) error {

	cronSpec := cronjob.Spec
	if !strings.HasPrefix(cronjob.Spec, "@") {
		cronSpec = fmt.Sprintf("0 %s", cronjob.Spec)
	}

	err := r.cron.AddFunc(cronSpec, r.cmdFunc(cronjob, func(execCmd *exec.Cmd) bool {
		// before exec callback
		log.WithFields(LogCronjobToFields(cronjob)).Debugf("executing")

		// lookup username
		u, err := user.Lookup(cronjob.User)
		if err != nil {
			log.WithFields(LogCronjobToFields(cronjob)).Errorf("user lookup failed: %v", err)
			return false
		}

		// convert userid to int
		userId, err := strconv.ParseUint(u.Uid, 10, 32)
		if err != nil {
			log.WithFields(LogCronjobToFields(cronjob)).Errorf("Cannot convert user to id:%v", err)
			return false
		}

		// convert groupid to int
		groupId, err := strconv.ParseUint(u.Gid, 10, 32)
		if err != nil {
			log.WithFields(LogCronjobToFields(cronjob)).Errorf("Cannot convert group to id:%v", err)
			return false
		}

		// add process credentials
		execCmd.SysProcAttr = &syscall.SysProcAttr{}
		execCmd.SysProcAttr.Credential = &syscall.Credential{Uid: uint32(userId), Gid: uint32(groupId)}
		return true
	}))

	if err != nil {
		prometheusMetricTask.With(r.cronjobToPrometheusLabels(cronjob)).Set(0)
		log.WithFields(LogCronjobToFields(cronjob)).Errorf("cronjob failed adding: %v", err)
	} else {
		prometheusMetricTask.With(r.cronjobToPrometheusLabels(cronjob)).Set(1)
		log.WithFields(LogCronjobToFields(cronjob)).Infof("cronjob added")
	}

	return err
}

// Return number of jobs
func (r *Runner) Len() int {
	return len(r.cron.Entries())
}

// Start runner
func (r *Runner) Start() {
	log.Infof("start runner with %d jobs\n", r.Len())
	r.cron.Start()
}

// Stop runner
func (r *Runner) Stop() {
	r.cron.Stop()
	log.Infof("stop runner")
}

// Execute crontab command
func (r *Runner) cmdFunc(cronjob CrontabEntry, cmdCallback func(*exec.Cmd) bool) func() {
	cmdFunc := func() {
		// fall back to normal shell if not specified
		taskShell := cronjob.Shell
		if taskShell == "" {
			taskShell = DEFAULT_SHELL
		}

		start := time.Now()

		// Init command
		execCmd := exec.Command(taskShell, "-c", cronjob.Command)

		// add custom env to cronjob
		if len(cronjob.Env) >= 1 {
			execCmd.Env = append(os.Environ(), cronjob.Env...)
		}

		// exec custom callback
		if cmdCallback(execCmd) {

			// exec job
			cmdStdout, err := execCmd.CombinedOutput()

			elapsed := time.Since(start)

			prometheusMetricTaskRunDuration.With(r.cronjobToPrometheusLabels(cronjob)).Set(elapsed.Seconds())
			prometheusMetricTaskRunTime.With(r.cronjobToPrometheusLabels(cronjob)).SetToCurrentTime()

			logFields := LogCronjobToFields(cronjob)
			if err != nil {
				prometheusMetricTaskRunCount.With(r.cronjobToPrometheusLabels(cronjob, prometheus.Labels{"result": "error"})).Inc()
				prometheusMetricTaskRunResult.With(r.cronjobToPrometheusLabels(cronjob)).Set(0)
				logFields["result"] = "error"
				log.WithFields(logFields).Errorln(string(cmdStdout))
			} else {
				prometheusMetricTaskRunCount.With(r.cronjobToPrometheusLabels(cronjob, prometheus.Labels{"result": "success"})).Inc()
				prometheusMetricTaskRunResult.With(r.cronjobToPrometheusLabels(cronjob)).Set(1)
				logFields["result"] = "success"
				log.WithFields(logFields).Debugln(string(cmdStdout))
			}
		}
	}
	return cmdFunc
}

func (r *Runner) cronjobToPrometheusLabels(cronjob CrontabEntry, additionalLabels ...prometheus.Labels) (labels prometheus.Labels) {
	labels = prometheus.Labels{
		"cronSpec":    cronjob.Spec,
		"cronUser":    cronjob.User,
		"cronCommand": cronjob.Command,
	}
	for _, additionalLabelValue := range additionalLabels {
		for labelName, labelValue := range additionalLabelValue {
			labels[labelName] = labelValue
		}
	}
	return
}


main.go
package main

import (
	"fmt"
	flags "github.com/jessevdk/go-flags"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	log "github.com/sirupsen/logrus"
	"net/http"
	"os"
	"os/signal"
	"os/user"
	"path/filepath"
	"runtime"
	"strings"
	"syscall"
)

const (
	Name                = "go-crond"
	Author              = "webdevops.io"
	CRONTAB_TYPE_SYSTEM = ""
)

var (
	// Git version information
	gitCommit = "<unknown>"
	gitTag    = "<unknown>"
)

var opts struct {
	DefaultUser         string   `long:"default-user"         description:"Default user"                  default:"root"`
	IncludeCronD        []string `long:"include"              description:"Include files in directory as system crontabs (with user)"`
	Auto                bool     `long:"auto"                 description:"Enable automatic system crontab detection"`
	RunParts            []string `long:"run-parts"            description:"Execute files in directory with custom spec (like run-parts; spec-units:ns,us,s,m,h; format:time-spec:path; eg:10s,1m,1h30m)"`
	RunParts1m          []string `long:"run-parts-1min"       description:"Execute files in directory every beginning minute (like run-parts)"`
	RunParts15m         []string `long:"run-parts-15min"      description:"Execute files in directory every beginning 15 minutes (like run-parts)"`
	RunPartsHourly      []string `long:"run-parts-hourly"     description:"Execute files in directory every beginning hour (like run-parts)"`
	RunPartsDaily       []string `long:"run-parts-daily"      description:"Execute files in directory every beginning day (like run-parts)"`
	RunPartsWeekly      []string `long:"run-parts-weekly"     description:"Execute files in directory every beginning week (like run-parts)"`
	RunPartsMonthly     []string `long:"run-parts-monthly"    description:"Execute files in directory every beginning month (like run-parts)"`
	AllowUnprivileged   bool     `long:"allow-unprivileged"   description:"Allow daemon to run as non root (unprivileged) user"`
	EnableUserSwitching bool
	ShowVersion         bool `short:"V"  long:"version"       description:"show version and exit"`
	ShowOnlyVersion     bool `long:"dumpversion"              description:"show only version number and exit"`
	ShowHelp            bool `short:"h"  long:"help"          description:"show this help message"`

	// logger
	Verbose bool `short:"v"  long:"verbose"      env:"VERBOSE"  description:"verbose mode"`
	LogJson bool `           long:"log.json"     env:"LOG_JSON" description:"Switch log output to json format"`

	// server settings
	ServerBind    string `long:"server.bind"     env:"SERVER_BIND"     description:"Server address, eg. ':8080' (/healthz and /metrics for prometheus)" default:""`
	ServerMetrics bool   `long:"server.metrics"  env:"SERVER_METRICS"  description:"Enable prometheus metrics (do not use senstive informations in commands -> use environment variables or files for storing these informations)"`
}

var argparser *flags.Parser
var args []string

func initArgParser() []string {
	var err error
	argparser = flags.NewParser(&opts, flags.PassDoubleDash)
	args, err = argparser.Parse()

	// check if there is an parse error
	if err != nil {
		logFatalErrorAndExit(err, 1)
	}

	// --dumpversion
	if opts.ShowOnlyVersion {
		fmt.Println(gitTag)
		os.Exit(0)
	}

	// --version
	if opts.ShowVersion {
		fmt.Println(fmt.Sprintf("%s version %s (%s)", Name, gitTag, gitCommit))
		fmt.Println(fmt.Sprintf("Copyright (C) 2020 %s", Author))
		os.Exit(0)
	}

	// --help
	if opts.ShowHelp {
		argparser.WriteHelp(os.Stdout)
		os.Exit(1)
	}

	// verbose level
	if opts.Verbose {
		log.SetLevel(log.DebugLevel)
	}

	// verbose level
	if opts.LogJson {
		log.SetFormatter(&log.JSONFormatter{})
	}

	return args
}

// Log error object as message
func logFatalErrorAndExit(err error, exitCode int) {
	if err != nil {
		log.Errorln(err)
	} else {
		log.Errorf("Unknown fatal error")
	}
	os.Exit(exitCode)
}

func findFilesInPaths(pathlist []string, callback func(os.FileInfo, string)) {
	for _, path := range pathlist {
		if stat, err := os.Stat(path); err == nil && stat.IsDir() {
			err := filepath.Walk(path, func(path string, f os.FileInfo, err error) error {
				path, _ = filepath.Abs(path)

				if f.IsDir() {
					return nil
				}

				if checkIfFileIsValid(f, path) {
					callback(f, path)
				}

				return nil
			})
			if err != nil {
				log.Fatal(err)
			}
		} else {
			log.Infof("path %s does not exists\n", path)
		}
	}
}

func findExecutabesInPathes(pathlist []string, callback func(os.FileInfo, string)) {
	findFilesInPaths(pathlist, func(f os.FileInfo, path string) {
		if f.Mode().IsRegular() && (f.Mode().Perm()&0100 != 0) {
			callback(f, path)
		} else {
			log.Infof("ignoring non exectuable file %s\n", path)
		}
	})
}

func includePathsForCrontabs(paths []string, username string) []CrontabEntry {
	var ret []CrontabEntry
	findFilesInPaths(paths, func(f os.FileInfo, path string) {
		entries := parseCrontab(path, username)
		ret = append(ret, entries...)
	})
	return ret
}

func includePathForCrontabs(path string, username string) []CrontabEntry {
	var ret []CrontabEntry
	var paths []string = []string{path}

	findFilesInPaths(paths, func(f os.FileInfo, path string) {
		entries := parseCrontab(path, username)
		ret = append(ret, entries...)
	})
	return ret
}

func includeRunPartsDirectories(spec string, paths []string) []CrontabEntry {
	var ret []CrontabEntry

	for _, path := range paths {
		ret = append(ret, includeRunPartsDirectory(spec, path)...)
	}

	return ret
}

func includeRunPartsDirectory(spec string, path string) []CrontabEntry {
	var ret []CrontabEntry

	user := opts.DefaultUser

	// extract user from path
	if strings.Contains(path, ":") {
		split := strings.SplitN(path, ":", 2)
		user, path = split[0], split[1]
	}

	var paths []string = []string{path}
	findExecutabesInPathes(paths, func(f os.FileInfo, path string) {
		ret = append(ret, CrontabEntry{Spec: spec, User: user, Command: path})
	})
	return ret
}

func parseCrontab(path string, username string) []CrontabEntry {
	var parser *Parser
	var err error

	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("crontab path: %v err:%v", path, err)
	}

	if username == CRONTAB_TYPE_SYSTEM {
		parser, err = NewCronjobSystemParser(file)
	} else {
		parser, err = NewCronjobUserParser(file, username)
	}

	if err != nil {
		log.Fatalf("parser read err: %v", err)
	}

	crontabEntries := parser.Parse()

	return crontabEntries
}

func collectCrontabs(args []string) []CrontabEntry {
	var ret []CrontabEntry

	// include system default crontab
	if opts.Auto {
		ret = append(ret, includeSystemDefaults()...)
	}

	// args: crontab files as normal arguments
	for _, crontabPath := range args {
		crontabUser := CRONTAB_TYPE_SYSTEM

		if strings.Contains(crontabPath, ":") {
			split := strings.SplitN(crontabPath, ":", 2)
			crontabUser, crontabPath = split[0], split[1]
		}

		crontabAbsPath, f := fileGetAbsolutePath(crontabPath)
		if checkIfFileIsValid(f, crontabAbsPath) {
			entries := parseCrontab(crontabAbsPath, crontabUser)
			ret = append(ret, entries...)
		}
	}

	// --include-crond
	if len(opts.IncludeCronD) >= 1 {
		ret = append(ret, includePathsForCrontabs(opts.IncludeCronD, CRONTAB_TYPE_SYSTEM)...)
	}

	// --run-parts
	if len(opts.RunParts) >= 1 {
		for _, runPart := range opts.RunParts {
			if strings.Contains(runPart, ":") {
				split := strings.SplitN(runPart, ":", 2)
				cronSpec, cronPath := split[0], split[1]
				cronSpec = fmt.Sprintf("@every %s", cronSpec)

				ret = append(ret, includeRunPartsDirectory(cronSpec, cronPath)...)
			} else {
				log.Infof("ignoring --run-parts because of missing time spec: %s\n", runPart)
			}
		}
	}

	// --run-parts-1min
	if len(opts.RunParts1m) >= 1 {
		ret = append(ret, includeRunPartsDirectories("@every 1m", opts.RunParts1m)...)
	}

	// --run-parts-15min
	if len(opts.RunParts15m) >= 1 {
		ret = append(ret, includeRunPartsDirectories("*/15 * * * *", opts.RunParts15m)...)
	}

	// --run-parts-hourly
	if len(opts.RunPartsHourly) >= 1 {
		ret = append(ret, includeRunPartsDirectories("@hourly", opts.RunPartsHourly)...)
	}

	// --run-parts-daily
	if len(opts.RunPartsDaily) >= 1 {
		ret = append(ret, includeRunPartsDirectories("@daily", opts.RunPartsDaily)...)
	}

	// --run-parts-weekly
	if len(opts.RunPartsWeekly) >= 1 {
		ret = append(ret, includeRunPartsDirectories("@weekly", opts.RunPartsWeekly)...)
	}

	// --run-parts-monthly
	if len(opts.RunPartsMonthly) >= 1 {
		ret = append(ret, includeRunPartsDirectories("@monthly", opts.RunPartsMonthly)...)
	}

	return ret
}

func includeSystemDefaults() []CrontabEntry {
	var ret []CrontabEntry

	systemDetected := false

	// ----------------------
	// Alpine
	// ----------------------
	if checkIfFileExistsAndOwnedByRoot("/etc/alpine-release") {
		log.Infof(" --> detected Alpine family, using distribution defaults")

		if checkIfDirectoryExists("/etc/crontabs") {
			ret = append(ret, includePathForCrontabs("/etc/crontabs", opts.DefaultUser)...)
		}

		systemDetected = true
	}

	// ----------------------
	// RedHat
	// ----------------------
	if checkIfFileExistsAndOwnedByRoot("/etc/redhat-release") {
		log.Infof(" --> detected RedHat family, using distribution defaults")

		if checkIfFileExists("/etc/crontabs") {
			ret = append(ret, includePathForCrontabs("/etc/crontabs", CRONTAB_TYPE_SYSTEM)...)
		}

		systemDetected = true
	}

	// ----------------------
	// SuSE
	// ----------------------
	if checkIfFileExistsAndOwnedByRoot("/etc/SuSE-release") {
		log.Infof(" --> detected SuSE family, using distribution defaults")

		if checkIfFileExists("/etc/crontab") {
			ret = append(ret, includePathForCrontabs("/etc/crontab", CRONTAB_TYPE_SYSTEM)...)
		}

		systemDetected = true
	}

	// ----------------------
	// Debian
	// ----------------------
	if checkIfFileExistsAndOwnedByRoot("/etc/debian_version") {
		log.Infof(" --> detected Debian family, using distribution defaults")

		if checkIfFileExists("/etc/crontab") {
			ret = append(ret, includePathForCrontabs("/etc/crontab", CRONTAB_TYPE_SYSTEM)...)
		}

		systemDetected = true
	}

	// ----------------------
	// General
	// ----------------------
	if !systemDetected {
		if checkIfFileExists("/etc/crontab") {
			ret = append(ret, includePathForCrontabs("/etc/crontab", CRONTAB_TYPE_SYSTEM)...)
		}

		if checkIfFileExists("/etc/crontabs") {
			ret = append(ret, includePathForCrontabs("/etc/crontabs", CRONTAB_TYPE_SYSTEM)...)
		}
	}

	if checkIfDirectoryExists("/etc/cron.d") {
		ret = append(ret, includePathForCrontabs("/etc/cron.d", CRONTAB_TYPE_SYSTEM)...)
	}

	return ret
}

func createCronRunner(args []string) *Runner {
	crontabEntries := collectCrontabs(args)

	runner := NewRunner()

	for _, crontabEntry := range crontabEntries {
		if opts.EnableUserSwitching {
			if err := runner.AddWithUser(crontabEntry); err != nil {
				log.Fatal(err)
			}
		} else {
			if err := runner.Add(crontabEntry); err != nil {
				log.Fatal(err)
			}
		}
	}

	return runner
}

func main() {
	args := initArgParser()

	c := make(chan os.Signal, 1)
	signal.Notify(c, syscall.SIGHUP)

	log.Infof("starting %s version %s (%s; %s) ", Name, gitTag, gitCommit, runtime.Version())

	// check if user switching is possible (have to be root)
	opts.EnableUserSwitching = true
	currentUser, err := user.Current()
	if err != nil || currentUser.Uid != "0" {
		if opts.AllowUnprivileged {
			log.Warnln("go-crond is NOT running as root, disabling user switching")
			opts.EnableUserSwitching = false
		} else {
			log.Errorln("go-crond is NOT running as root, add option --allow-unprivileged if this is ok")
			os.Exit(1)
		}
	}

	// get current path
	confDir, err := os.Getwd()
	if err != nil {
		log.Fatalf("could not get current path: %v", err)
	}

	// daemon mode
	initMetrics()
	log.Infof("starting http server on %s", opts.ServerBind)
	startHttpServer()

	// endless daemon-reload loop
	for {
		resetMetrics()

		// change to initial directory for fetching crontabs
		err = os.Chdir(confDir)
		if err != nil {
			log.Fatalf("cannot switch to path %s: %v", confDir, err)
		}

		// create new cron runner
		runner := createCronRunner(args)
		registerRunnerShutdown(runner)

		// chdir to root to prevent relative path errors
		err = os.Chdir("/")
		if err != nil {
			log.Fatalf("cannot switch to path %s: %v", confDir, err)
		}

		// start new cron runner
		runner.Start()

		// check if we received SIGHUP and start a new loop
		s := <-c
		log.Infof("Got signal: %v", s)
		runner.Stop()
		log.Infof("Reloading configuration")
	}
}

// start and handle prometheus handler
func startHttpServer() {
	go func() {
		if opts.ServerBind != "" {
			// healthz
			http.HandleFunc("/healthz", func(w http.ResponseWriter, r *http.Request) {
				if _, err := fmt.Fprint(w, "Ok"); err != nil {
					log.Errorf(err.Error())
				}
			})

			if opts.ServerMetrics {
				http.Handle("/metrics", promhttp.Handler())
			}
			if err := http.ListenAndServe(opts.ServerBind, nil); err != nil {
				panic(err)
			}
		}
	}()
}

func registerRunnerShutdown(runner *Runner) {
	c := make(chan os.Signal, 2)
	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
	go func() {
		s := <-c
		log.Infof("got signal: %v", s)
		runner.Stop()

		log.Infof("terminated")
		os.Exit(1)
	}()
}
